designer/server/src/main/scala/pl/touk/nussknacker/ui/api/app/AppApiHttpService.scala

```diff
@@ -66,43 +58,43 @@ class AppApiHttpService(config: Config,
               } else {
                 logger.warn(s"Scenarios with status PROBLEM: ${set.keys}")
                 logger.debug(s"Scenarios with status PROBLEM: $set")
-                Left(HealthCheckProcessErrorResponseDto(
+                Left(Left(HealthCheckProcessErrorResponseDto(
```

Comment:
Tapir has only one type parameter for errors (see [the github issue](https://github.com/softwaremill/tapir/issues/2654)). It means that when you want to split Security-based errors and Business-related errors you have two ways:
1. GADT where the business error type will be provided using a generic type
2. `Either[SecurityError, BusinessError]`

I've checked both and it seems that in the case of the current version of Tapir, the second approach is more seamless. Even that we have the two Lefts here. 

But you are right in general - I'll provide some util to hide it a little bit.

Reply:
IMO GADT would be more readable. What exactly is more seamless?

Reply:
It was my first choice, but sadly it doesn't fit well with Tapir's approach. As you can see in the link, even the author of the lib suggests using `Either`.

WDYT about my changes related to these two Lefts?

Reply:
about the "seamlessness": 
When have a separate type for security errors you can describe the part of the endpoint separately with ease:
```

    def withSecurity(auth: Auth[AuthCredentials, _]): SecuredEndpoint[INPUT, BUSINESS_ERROR, OUTPUT, R] = {
      endpoint
        .securityIn(auth)
        .errorOutEither(
          oneOf(
            oneOfVariantFromMatchType(Unauthorized, plainBody[SecurityError.AuthenticationError.type]),
            oneOfVariantFromMatchType(Forbidden, plainBody[SecurityError.AuthorizationError.type])
          )
        )
    }    
```

Then in case of specific endpoint you deal with the business error only - like this:

```
    baseNuApiEndpoint
      .get
      .in("app" / "healthCheck" / "process" / "validation")
      .out(
        statusCode(Ok).and(jsonBody[HealthCheckProcessSuccessResponseDto.type])
      )
      .errorOut(
        statusCode(InternalServerError).and(jsonBody[HealthCheckProcessErrorResponseDto])
      )
      .withSecurity(auth)
```

(see the `errorOutXXX` methods).

In the case of GADT, it's really huge pain in the ass to do it.

Reply:
I understand. WDYT about some little tweaks that will make this solution even more readable:
1. Add the third method `securityError` next to existing
2. Add type alias e.g. `LogicResult[BusinessError, Result] = Either[Either[BusinessError, SecurityError], Result]`
2. a) Or at least type alias for `LogicError[BusinessError] = Either[BusinessError, SecurityError]`
3. Add the link to this issue next to these methods and types

Thanks to that IMO will be easier to figure out in which place in the type is given information

Reply:
nice thanks